### PR TITLE
Liveness analysis

### PR DESCRIPTION
**Pull request by bebidek**
Some changes in Instruction.kt are temporary and are meant to be overwritten by "Instructions" PR.

### CODE REVIEW

File: app/src/main/kotlin/compiler/intermediate_form/DataFlowAnalyser.kt [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        return referenceMapOf(predecessors.map { it.toPair() })
+    }
+
+    private fun getEntryPoints(instructionList: List<Instruction>): ReferenceSet<Instruction> {
+        return if (backward)
+            referenceHashSetOf(instructionList.filterIsInstance<Instruction.RetInstruction>())
+        else
+            referenceHashSetOf(instructionList.first())
```
Comment (JHodorHodor) [inline app/src/main/kotlin/compiler/intermediate_form/DataFlowAnalyser.kt:66]: Is `instructionList` empty for the empty program?

File: app/src/main/kotlin/compiler/intermediate_form/DataFlowAnalyser.kt [hunk 2]
```diff
@@ -0,0 +1,110 @@
+package compiler.intermediate_form
+
+import compiler.common.reference_collections.ReferenceMap
+import compiler.common.reference_collections.ReferenceSet
+import compiler.common.reference_collections.referenceHashMapOf
+import compiler.common.reference_collections.referenceHashSetOf
+import compiler.common.reference_collections.referenceMapOf
+
+abstract class DataFlowAnalyser<SL> {
```
Comment (JHodorHodor) [inline app/src/main/kotlin/compiler/intermediate_form/DataFlowAnalyser.kt:9]:
```suggestion
abstract class DataFlowAnalyzer<SL> {
```
let's be consistent ;)

File: app/src/main/kotlin/compiler/intermediate_form/DataFlowAnalyser.kt [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    private fun getEntryPoints(instructionList: List<Instruction>): ReferenceSet<Instruction> {
+        return if (backward)
+            referenceHashSetOf(instructionList.filterIsInstance<Instruction.RetInstruction>())
+        else
+            referenceHashSetOf(instructionList.first())
+    }
+
+    data class DataFlowResult<SL>(val inValues: ReferenceMap<Instruction, SL>, val outValues: ReferenceMap<Instruction, SL>)
+
+    fun analyse(linearProgram: List<Asmable>): DataFlowResult<SL> {
```
Comment (JHodorHodor) [inline app/src/main/kotlin/compiler/intermediate_form/DataFlowAnalyser.kt:71]:
```suggestion
    fun analyze(linearProgram: List<Asmable>): DataFlowResult<SL> {
```

File: app/src/main/kotlin/compiler/intermediate_form/Liveness.kt [hunk 4]
```diff
@@ -1,10 +1,65 @@
...
+
+        // run analysis and fill the graphs
+        val dataFlowResult = LivenessDataFlowAnalyser.analyse(linearProgram)
+        val outLiveRegisters = dataFlowResult.outValues.mapValues { (instr, regs) -> combineReferenceSets(regs, referenceHashSetOf(instr.regsDefined.toList())) }
+
+        instructionList.forEach {
+            for (definedReg in it.regsDefined)
+                for (regLiveOnOutput in outLiveRegisters[it]!!)
+                    if (definedReg !== regLiveOnOutput) {
+                        if (it is Instruction.InPlaceInstruction.MoveRR && it.reg_dest == definedReg && it.reg_src == regLiveOnOutput) {
```
Comment (JHodorHodor) [inline app/src/main/kotlin/compiler/intermediate_form/Liveness.kt:53]: I'm not 100% sure, but if the instruction is `x := y`, and `y` is never used again, then y is not in `outLiveRegisters[it]!!`, and thus there is no `x--y` edge in the copy graph, but we would like it to be there.

File: app/src/test/kotlin/compiler/intermediate_form/LivenessTest.kt [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+        val livenessGraphs = Liveness.computeLiveness(linearProgram)
+        assertEquals(expectedLivenessGraphs, livenessGraphs)
+    }
+
+    @Test
+    fun `copying and interference`() {
+        val linearProgram = listOf(
+            Instruction.InPlaceInstruction.Dummy(regsDefined = listOf(reg1, reg2, reg3)),
+            Instruction.InPlaceInstruction.MoveRR(reg2, reg1),
+            Instruction.InPlaceInstruction.Dummy(regsUsed = listOf(reg1, reg2, reg3)),
```
Comment (JHodorHodor) [inline app/src/test/kotlin/compiler/intermediate_form/LivenessTest.kt:244]: According to my comment above. If I understand correctly, the removal of `reg1` in this line shouldn't change the result and it changes.

**Review by JHodorHodor — CHANGES_REQUESTED** [review]
Mam pewną wątpliwość, czy program działa tak jak powinien, zostawiłem w komentarzach.

File: app/src/main/kotlin/compiler/intermediate_form/DataFlowAnalyser.kt (hunk 1 above)
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/DataFlowAnalyser.kt:66]: I assume that every execution path must lead to RET instruction.

File: app/src/main/kotlin/compiler/intermediate_form/Liveness.kt (hunk 4 above)
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/Liveness.kt:53]: You're right, corrected and added a test for that.

**Review by JHodorHodor — APPROVED** [review]

**Pull request merged**